cloud/azure/azure_deployment.py

```diff
@@ -0,0 +1,620 @@
+        if e.status_code == 404 or e.status_code == 204:
+            return True
+        else:
+            module.fail_json(
+                msg='Delete resource group and deploy failed with status code: %s and message: %s' % (e.status_code, e.message))
+
+
+def get_dependencies(dep_tree, resource_type):
+    matches = [value for value in dep_tree.values() if value['dep'].resource_type == resource_type]
+    for child_tree in [value['children'] for value in dep_tree.values()]:
+        matches += get_dependencies(child_tree, resource_type)
+    return matches
+
+
+def build_hierarchy(dependencies, tree=None):
+    tree = dict(top=True) if tree is None else tree
+    for dep in dependencies:
+        if dep.resource_name not in tree:
+            tree[dep.resource_name] = dict(dep=dep, children=dict())
+        if isinstance(dep, Dependency) and dep.depends_on is not None and len(dep.depends_on) > 0:
+            build_hierarchy(dep.depends_on, tree[dep.resource_name]['children'])
+
+    if 'top' in tree:
+        tree.pop('top', None)
+        keys = list(tree.keys())
+        for key1 in keys:
+            for key2 in keys:
+                if key2 in tree and key1 in tree[key2]['children'] and key1 in tree:
+                    tree[key2]['children'][key1] = tree[key1]
+                    tree.pop(key1)
+    return tree
+
+
+def get_ip_dict(ip):
+    ip_dict = dict(name=ip.name,
+                id=ip.id,
+                public_ip=ip.ip_address,
+                public_ip_allocation_method=str(ip.public_ip_allocation_method))
+
+    if ip.dns_settings:
+        ip_dict['dns_settings'] = {
+                    'domain_name_label':ip.dns_settings.domain_name_label,
+                    'fqdn':ip.dns_settings.fqdn
+        }
+
+    return ip_dict
+
+
+def nic_to_public_ips_instance(client, group, nics):
+    return [client.public_ip_addresses.get(group, public_ip_id.split('/')[-1])
+              for nic_obj in [client.network_interfaces.get(group, nic['dep'].resource_name) for nic in nics]
+              for public_ip_id in [ip_conf_instance.public_ip_address.id for ip_conf_instance in nic_obj.ip_configurations if ip_conf_instance.public_ip_address]]
+
+
+def get_instances(client, group, deployment):
+    dep_tree = build_hierarchy(deployment.properties.dependencies)
+    vms = get_dependencies(dep_tree, resource_type="Microsoft.Compute/virtualMachines")
+
+    vms_and_nics = [(vm, get_dependencies(vm['children'], "Microsoft.Network/networkInterfaces")) for vm in vms]
+    vms_and_ips = [(vm['dep'], nic_to_public_ips_instance(client, group, nics)) for vm, nics in vms_and_nics]
+
+    return [dict(vm_name=vm.resource_name, ips=[get_ip_dict(ip) for ip in ips]) for vm, ips in vms_and_ips if len(ips) > 0]
+
+
+def main():
+    argument_spec = dict(
+        azure_url=dict(default=AZURE_URL),
+        subscription_id=dict(),
+        client_secret=dict(no_log=True),
+        client_id=dict(required=True),
+        tenant_id=dict(required=True),
```

Comment:
I wouldn't worry about it much- that stuff will all change when this gets merged anyway. We've got shared module_utils code for the other Azure RM modules that does all this stuff (as well as defining an on-disk credential storage similar to the way boto/aws work).